src/welcomePage/WelcomePageOnboarding/onboardingContext/utils.ts

```diff
@@ -54,29 +53,41 @@ export const getFirstExpanderStepId = (onboardingState: OnboardingState) => {
   return (stepsExpandedEntries.find(([_, value]) => value)?.[0] ?? "") as OnboardingStepsIDs;
 };
 
-export const mapInitialStepsWithState = (onboardingState: OnboardingState) =>
-  [...initialOnboardingSteps].map(step => ({
+export const getFirstNotCompletedAndNotExpandedStep = (
+  onboardingState: OnboardingState,
+  visibleSteps: OnboardingStep[],
+): OnboardingStepsIDs | "" => {
+  const stepsWithState = visibleSteps.map(step => ({
     ...step,
     completed: onboardingState.stepsCompleted.includes(step.id),
     expanded: onboardingState.stepsExpanded[step.id],
   }));
 
-export const getFirstNotCompletedAndNotExpandedStep = (onboardingState: OnboardingState) => {
-  return (
-    mapInitialStepsWithState(onboardingState).filter(
-      step => !step.completed && step.expanded !== false,
-    )[0]?.id ?? ""
-  );
+  return stepsWithState.find(step => !step.completed && step.expanded !== false)?.id ?? "";
 };
 
-export const getNextStepToExpand = (onboardingState: OnboardingState) => {
+export const getNextStepToExpand = (
+  onboardingState: OnboardingState,
+  visibleSteps: OnboardingStep[],
+): OnboardingStepsIDs | "" => {
   const lastCompletedStepId =
     onboardingState.stepsCompleted[onboardingState.stepsCompleted.length - 1];
-  const steps = mapInitialStepsWithState(onboardingState);
-  const stepIndex = steps.findIndex(step => step.id === lastCompletedStepId);
+
+  const stepsWithState = visibleSteps.map(step => ({
+    ...step,
+    completed: onboardingState.stepsCompleted.includes(step.id),
+    expanded: onboardingState.stepsExpanded[step.id],
+  }));
+
+  const stepIndex = stepsWithState.findIndex(step => step.id === lastCompletedStepId);
+
+  if (stepIndex === -1 || stepIndex === stepsWithState.length - 1) {
+    return "";
+  }
 
   return (
-    steps.slice(stepIndex + 1).find(step => !step.completed && step.expanded !== false)?.id ?? ""
+    stepsWithState.slice(stepIndex + 1).find(step => !step.completed && step.expanded !== false)
```

Comment:
Tip: if you see something like that or do it yourself, it's a way better and more readable to delegate callback logic to the separate function and name it properly, example:

```
const arr = [1, 2, null, 3, 4, 5];

const toMultiplied = (a: number, b: number) => {
  return a * b;
};

const byNotNull = (a: number) => {
  return a !== null;
};

const byIsOdd = (a: number) => {
  return a % 2 !== 0;
};


const result = arr
  .filter(byNotNull)
  .filter(byIsOdd)
  .map(toMultiplied);
```

Now i can basically read this almost as sentence in english + worth keep this in a separate file for keeping separation